sycl/source/builtins_generator.py

```diff
@@ -0,0 +1,1285 @@
+  # are used and indicate the index of the used element.
+  def convert_loop_arg(self, arg_type, arg_name):
+    if isinstance(arg_type, MultiPtr):
+      return f'address_space_cast<Space, IsDecorated, detail::get_elem_type_t<{arg_type.element_type}>>(&(*{arg_name})[I])'
+    if is_marray_arg(arg_type):
+      return str(arg_name) + '[I]'
+    return str(arg_name)
+
+  # Generated the body of a loop-based implementation of marray builtins.
+  def get_marray_loop_invoke_body(self, namespaced_builtin_name, return_type,
+                                  arg_types, arg_names, first_marray_type):
+    result = ""
+    args = [self.convert_loop_arg(arg_type, arg_name) for arg_type, arg_name in zip(arg_types, arg_names)]
+    joined_args = ', '.join(args)
+    (size_alias, size_alias_init) = self.require_size_alias('N', first_marray_type)
+    result += size_alias_init
+    return result + f"""  {return_type} Res;
+  for (int I = 0; I < {size_alias}; ++I)
+    Res[I] = {namespaced_builtin_name}({joined_args});
+  return Res;"""
+
+  # Generates the body of a direct vector cast implementation for marray
+  # builtins.
+  def get_marray_vec_cast_invoke_body(self, namespaced_builtin_name,
+                                      return_type, arg_types, arg_names,
+                                      first_marray_type):
+    result = ""
+    vec_cast_args = [f'detail::to_vec({arg_name})' if is_marray_arg(arg_type) else str(arg_name) for arg_type, arg_name in zip(arg_types, arg_names)]
+    joined_vec_cast_args = ', '.join(vec_cast_args)
+    vec_call = f'{namespaced_builtin_name}({joined_vec_cast_args})'
+    if isinstance(return_type, InstantiatedTemplatedReturnType):
+      # Convert the vec call result to marray.
+      vec_call = f'detail::to_marray({vec_call})'
+    return result + f'  return {vec_call};'
+
+  # Generates the body of a vectorized implementation for marray builtins.
+  def get_marray_vectorized_invoke_body(self, namespaced_builtin_name,
+                                        return_type, arg_types, arg_names,
+                                        first_marray_type):
+    result = ""
+    (size_alias, size_alias_init) = self.require_size_alias('N', first_marray_type)
+    result += size_alias_init
+    # Adjust arguments for partial results and the remaining work at the end.
+    imm_args = []
+    rem_args = []
+    for arg_type, arg_name in zip(arg_types, arg_names):
+      is_marray = is_marray_arg(arg_type)
+      imm_args.append(f'detail::to_vec2({arg_name}, I * 2)' if is_marray else arg_name)
+      rem_args.append(f'{arg_name}[{size_alias} - 1]' if is_marray else arg_name)
+    joined_imm_args = ', '.join(imm_args)
+    joined_rem_args = ', '.join(rem_args)
+    return result + f"""  {return_type} Res;
+  for (size_t I = 0; I < {size_alias} / 2; ++I) {{
+    auto PartialRes = {namespaced_builtin_name}({joined_imm_args});
+    std::memcpy(&Res[I * 2], &PartialRes, sizeof(decltype(PartialRes)));
```

Comment:
Note: this code uses non-standard SYCL functionality, there is no guarantee that `std::memcpy` will work on device.

FYI: a few years ago we even had a problem with `std::memcpy` in some build environments: intel/llvm#1117

Reply:
Agreed. We may want to revisit how these are implemented.